### PR TITLE
End to end test rule concurrency

### PR DESCRIPTION
**Pull request by gotjosh**
#### What this PR does

Adds an end to ends test to ensure rule concurrency is working correctly. A follow up to https://github.com/grafana/mimir/pull/8146

#### Which issue(s) this PR fixes or relates to

N/A

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [x] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: integration/ruler_test.go [hunk 1]
```diff
@@ -1478,6 +1478,62 @@ func TestRulerProtectedNamespaces(t *testing.T) {
...
+	}
+
+	require.NoError(t, c.SetRuleGroup(rg, "test"))
+	require.NoError(t, c2.SetRuleGroup(rg, "test"))
+
+	// Wait until rulers have loaded all rules.
+	require.NoError(t, ruler.WaitSumMetricsWithOptions(e2e.Equals(20), []string{"cortex_prometheus_rule_group_rules"}, e2e.WaitMissingMetrics))
+
+	// We should have at least 20 attempts.
+	require.NoError(t, ruler.WaitSumMetricsWithOptions(e2e.Equals(20), []string{"cortex_ruler_independent_rule_evaluation_concurrency_attempts_started_total"}, e2e.WaitMissingMetrics))
```
Comment (gotjosh) [inline integration/ruler_test.go:1534]: I don't think there's anything else I can test reliably - I could introduce superfluous checks that are `<= currently in use` but I don't think this adds any value?

File: integration/ruler_test.go [hunk 2]
```diff
@@ -1478,6 +1478,70 @@ func TestRulerProtectedNamespaces(t *testing.T) {
...
+	// Wait until rulers have loaded all rules.
+	require.NoError(t, ruler.WaitSumMetricsWithOptions(e2e.Equals(20), []string{"cortex_prometheus_rule_group_rules"}, e2e.WaitMissingMetrics))
+
+	// We should have 20 attempts and 20 or less for failed or successful attempts to acquire the lock.
+	require.NoError(t, ruler.WaitSumMetricsWithOptions(e2e.Equals(20), []string{"cortex_ruler_independent_rule_evaluation_concurrency_attempts_started_total"}, e2e.WaitMissingMetrics))
+	require.NoError(t, ruler.WaitSumMetricsWithOptions(e2e.Less(21), []string{"cortex_ruler_independent_rule_evaluation_concurrency_attempts_completed_total"}, e2e.WaitMissingMetrics))
+	require.NoError(t, ruler.WaitSumMetricsWithOptions(e2e.Less(21), []string{"cortex_ruler_independent_rule_evaluation_concurrency_attempts_incomplete_total"}, e2e.WaitMissingMetrics))
+
+	// We should never have more than the allowed concurrency slots at the same time.
+	require.NoError(t, ruler.WaitSumMetricsWithOptions(e2e.Less(4), []string{"cortex_ruler_independent_rule_evaluation_concurrency_slots_in_use"}, e2e.WaitMissingMetrics))
```
Comment (pracucci) [inline integration/ruler_test.go:1542]: I don't think this assertion is much helpful. We're just testing it at a specific point in time, but at this time all rule evaluations may have completed and slot in use may be 0.

**Review by pracucci — APPROVED** [review]
Approving to unblock you, but I'm not much convinced the assertions increase our confidence.

File: integration/ruler_test.go [hunk 3]
```diff
@@ -1478,6 +1478,70 @@ func TestRulerProtectedNamespaces(t *testing.T) {
...
+
+	require.NoError(t, c.SetRuleGroup(rg, "fileY"))
+	require.NoError(t, c2.SetRuleGroup(rg, "fileY"))
+
+	// Wait until rulers have loaded all rules.
+	require.NoError(t, ruler.WaitSumMetricsWithOptions(e2e.Equals(20), []string{"cortex_prometheus_rule_group_rules"}, e2e.WaitMissingMetrics))
+
+	// We should have 20 attempts and 20 or less for failed or successful attempts to acquire the lock.
+	require.NoError(t, ruler.WaitSumMetricsWithOptions(e2e.Equals(20), []string{"cortex_ruler_independent_rule_evaluation_concurrency_attempts_started_total"}, e2e.WaitMissingMetrics))
+	require.NoError(t, ruler.WaitSumMetricsWithOptions(e2e.Less(21), []string{"cortex_ruler_independent_rule_evaluation_concurrency_attempts_completed_total"}, e2e.WaitMissingMetrics))
```
[the remark this answers is not in the document]
Comment (gotjosh) [reply integration/ruler_test.go]: I wasn't sure if you thought both were both implemented but I did both in https://github.com/grafana/mimir/pull/8846/commits/1efeb015476a2a3e8f75f8729d1dbc83c4ac28ff

File: integration/ruler_test.go (hunk 2 above)
Comment (gotjosh) [reply integration/ruler_test.go]: I agree - I ended up just removing it.

File: integration/ruler_test.go [hunk 4]
```diff
@@ -1478,6 +1478,75 @@ func TestRulerProtectedNamespaces(t *testing.T) {
...
+	require.NoError(t, ruler.WaitSumMetricsWithOptions(e2e.Equals(20), []string{"cortex_ruler_independent_rule_evaluation_concurrency_attempts_started_total"}, e2e.WaitMissingMetrics))
+	// The magic number here is because we have a maximum per tenant concurrency of 2. So we expect at least 4 (2 slots * 2 tenants) to complete successfully.
+	require.NoError(t, ruler.WaitSumMetricsWithOptions(e2e.GreaterOrEqual(4), []string{"cortex_ruler_independent_rule_evaluation_concurrency_attempts_completed_total"}, e2e.WaitMissingMetrics))
+	require.NoError(t, ruler.WaitSumMetricsWithOptions(func(sums ...float64) bool {
+		var total float64
+		for _, sum := range sums {
+			total += sum
+		}
+
+		return total == 20
```
Comment (pracucci) [inline integration/ruler_test.go:1546]: [super nit] I think can be simplified as follows:

```suggestion
		return e2e.SumValues(sums) == 20
```

**Review by pracucci — APPROVED** [review]
Nice, LGTM!

File: integration/ruler_test.go (hunk 4 above)
Comment (gotjosh) [reply integration/ruler_test.go:1546]: 🤦 DUH - I found that function and was wondering why it was there haha when is not a compatible `e2e.Result`.

Comment (gotjosh) [reply integration/ruler_test.go]: done in https://github.com/grafana/mimir/pull/8846/commits/3c9e144217ecab28d0459149f361d63137d49301

**Pull request merged**